### PR TITLE
Implemented couchdb scripts

### PR DESCRIPTION
**Pull request by akhilmhdh**
# Closes
1. 1207

# Explain the feature/fix

1. Implemented couchdb script to load couchdb with design documents and default tags
2. Changed pouchdb change event listening to onMount and avoided stale prop to save network calls and some performance boost
3. Added validation in idea creation form with the controller component
4. Added replication filter to avoid design documents getting replication to remote and added validation doc to validate documents in couchdb

## Does this PR introduce a breaking change

No

## Screenshots

Nil

### Ready-for-merge Checklist

- [x] Expected files: all files in this pull request are related to one feature request or issue (no stragglers)?
- [ ] Does the change have appropriate unit tests?
- [ ] Did tests pass?
- [x] Did you update or add any necessary documentation (README.md, WHY.md, etc.)?
- [x] Was this feature demo'd and the design review approved?

### CODE REVIEW

**Review by ghanlohar — CHANGES_REQUESTED** [review]

File: scripts/remote-pouch-db-instance.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import PouchDB from 'pouchdb';
+import PouchDBFind from 'pouchdb-find';
+import pouchdbHTTPAdapter from 'pouchdb-adapter-http';
+import dotenv from 'dotenv';
+dotenv.config();
+
+import { POUCHDB_DB_URL } from '../src/pouchDB/config';
+
+PouchDB.plugin(PouchDBFind);
+PouchDB.plugin(pouchdbHTTPAdapter);
```
Comment (deshmukhmayur) [inline scripts/remote-pouch-db-instance.ts:10]: The http adapter is not required, so you can remove it.

File: package.json [hunk 2]
```diff
@@ -20,10 +20,12 @@
     "@types/node": "^15.12.2",
     "@types/react": "^17.0.11",
     "@types/react-dom": "^17.0.7",
+    "dotenv": "^10.0.0",
     "history": "^5.0.0",
     "nanoid": "^3.1.23",
     "node-sass": "^6.0.1",
     "pouchdb": "^7.2.2",
```
Comment (deshmukhmayur) [inline package.json:27]: Also, replace `pouchdb` with `pouchdb-browser`.

**Review by deshmukhmayur — CHANGES_REQUESTED** [review]

File: src/pouchDB/design.ts [hunk 3]
```diff
@@ -74,28 +103,33 @@ const designDocuments: PouchDB.Core.PutDocument<{ version: number }>[] = [
...
+        const res = await db.put(designDocument);
+        console.log(res);
+      } catch (error) {
+        if (error.status === 409) {
+          const existingOneId = error.docId;
+          const exisitingDesignDoc = await db.get<{ version: number }>(existingOneId);
+          if (exisitingDesignDoc?.version < designDocument.version) {
+            designDocument._rev = exisitingDesignDoc._rev;
+            const res = await db.put(designDocument);
+            console.log(res);
```
[the remark this answers is not in the document]
Comment (akhilmhdh) [reply src/pouchDB/design.ts]: Done

File: src/pouchDB/design.ts [hunk 4]
```diff
@@ -74,28 +103,33 @@ const designDocuments: PouchDB.Core.PutDocument<{ version: number }>[] = [
...
-        const exisitingDesignDoc = await db.get<{ version: number }>(existingOneId);
-        if (exisitingDesignDoc?.version < designDocument.version) {
-          designDocument._rev = exisitingDesignDoc._rev;
-          const res = await db.put(designDocument);
-          console.log(res);
+  await Promise.all(
+    designDocuments.map(async (designDocument) => {
+      try {
+        const res = await db.put(designDocument);
+        console.log(res);
```
[the remark this answers is not in the document]
Comment (akhilmhdh) [reply src/pouchDB/design.ts]: Done

File: README.md [hunk 5]
```diff
@@ -30,6 +30,14 @@ run `npm test`
 
 run `npm run lint:fix`
 
+## Load remote couchdb with design documents and default tag
+
+run `npm run generate:couchdb-documents`: To save both design documents and default tags to couchdb
+
+run `nom run generate:design-doc`: To save design documents to couchdb
+
+run `nom run generate:default-tags`: To save default tags to couchdb
```
[the remark this answers is not in the document]
Comment (akhilmhdh) [reply README.md]: Done

File: package.json [hunk 6]
```diff
@@ -42,7 +44,10 @@
     "eject": "react-scripts eject",
     "lint": "eslint --ext .ts,.tsx src/",
     "lint:fix": "eslint --ext .ts,.tsx src/ --fix",
-    "prepare": "husky install"
+    "prepare": "husky install",
+    "generate:design-doc": "ts-node -P scripts/tsconfig.json scripts/generate-couchdb-design-documents.ts",
+    "generate:default-tags": "ts-node -P scripts/tsconfig.json scripts/generate-couchdb-default-tags.ts",
+    "generate:couchdb-documents": "npm run generate:design-doc && npm run generate:default-tags"
```
[the remark this answers is not in the document]
Comment (akhilmhdh) [reply package.json]: Done

File: package.json [hunk 7]
```diff
@@ -20,10 +20,12 @@
     "@types/node": "^15.12.2",
     "@types/react": "^17.0.11",
     "@types/react-dom": "^17.0.7",
+    "dotenv": "^10.0.0",
```
[the remark this answers is not in the document]
Comment (akhilmhdh) [reply package.json:24]: Done

File: scripts/remote-pouch-db-instance.ts (hunk 1 above)
Comment (akhilmhdh) [reply scripts/remote-pouch-db-instance.ts]: Removed

File: package.json (hunk 2 above)
Comment (akhilmhdh) [reply package.json]: Changed

**Review by ghanlohar — DISMISSED** [review]

File: README.md [hunk 8]
```diff
@@ -30,6 +31,14 @@ run `npm test`
 
 run `npm run lint:fix`
 
+## Load remote couchdb with design documents and default tag
+
+run `npm run migrate:couchdb-documents`: To save both design documents and default tags to couchdb
```
Comment (deshmukhmayur) [inline README.md:36]: This script is named `migrate:all` in the package.json. Keep it `:all`. That is better than the longer `couchdb-documents`
```suggestion
run `npm run migrate:all`: To save both design documents and default tags to couchdb
```

**Review by deshmukhmayur — CHANGES_REQUESTED** [review]

File: README.md (hunk 8 above)
Comment (akhilmhdh) [reply README.md:36]: My bad i forgot to change that..

Comment (akhilmhdh) [reply README.md]: Done

**Review by deshmukhmayur — APPROVED** [review]

**Comment by akhilmhdh** [comment]
@ghanlohar  and @riginoommen  Can you guys please review this

**Comment by akhilmhdh** [comment]
Mergin this PR as i got approval from all three/

CC: @ghanlohar  @riginoommen  @deshmukhmayur

**Pull request merged**